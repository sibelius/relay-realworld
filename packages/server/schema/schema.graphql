"""The root of all... queries"""
type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Fetches objects given their IDs"""
  nodes(
    """The IDs of objects"""
    ids: [ID!]!
  ): [Node]!
  me: User
  posts(after: String, first: Int, before: String, last: Int): PostConnection!
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!

  """mongoose _id"""
  _id: String!
  name: String
  email: String
  posts(after: String, first: Int, before: String, last: Int): PostConnection!
  comments(after: String, first: Int, before: String, last: Int): CommentConnection!
  createdAt: String
  updatedAt: String
}

"""A connection to a list of items."""
type PostConnection {
  """Number of items in this connection"""
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
    This allows a client to fetch the first five objects by passing "5" as the
    argument to "first", then fetch the total count so it could display "5 of 83",
    for example.
  """
  totalCount: Int

  """Offset from start"""
  startCursorOffset: Int!

  """Offset till end"""
  endCursorOffset: Int!

  """Information to aid in pagination."""
  pageInfo: PageInfoExtended!

  """A list of edges."""
  edges: [PostEdge]!
}

"""Information about pagination in a connection."""
type PageInfoExtended {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""An edge in a connection."""
type PostEdge {
  """The item at the end of the edge"""
  node: Post

  """A cursor for use in pagination"""
  cursor: String!
}

"""Post data"""
type Post implements Node {
  """The ID of an object"""
  id: ID!

  """mongoose _id"""
  _id: String!
  content: String
  author: User
  likesCount: Int!
  commentsCount: Int!
  comments(after: String, first: Int, before: String, last: Int): CommentConnection!

  """whether logged user liked this post"""
  meHasLiked: Boolean!
  createdAt: String
  updatedAt: String
}

"""A connection to a list of items."""
type CommentConnection {
  """Number of items in this connection"""
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
    This allows a client to fetch the first five objects by passing "5" as the
    argument to "first", then fetch the total count so it could display "5 of 83",
    for example.
  """
  totalCount: Int

  """Offset from start"""
  startCursorOffset: Int!

  """Offset till end"""
  endCursorOffset: Int!

  """Information to aid in pagination."""
  pageInfo: PageInfoExtended!

  """A list of edges."""
  edges: [CommentEdge]!
}

"""An edge in a connection."""
type CommentEdge {
  """The item at the end of the edge"""
  node: Comment

  """A cursor for use in pagination"""
  cursor: String!
}

"""Comment data"""
type Comment implements Node {
  """The ID of an object"""
  id: ID!

  """mongoose _id"""
  _id: String!
  body: String
  user: User
  post: Post
  likesCount: Int!

  """whether logged user liked this post"""
  meHasLiked: Boolean!
  createdAt: String
  updatedAt: String
}

type Mutation {
  UserLoginWithEmail(input: UserLoginWithEmailInput!): UserLoginWithEmailPayload
  UserRegisterWithEmail(input: UserRegisterWithEmailInput!): UserRegisterWithEmailPayload
  PostCreate(input: PostCreateInput!): PostCreatePayload
  PostLike(input: PostLikeInput!): PostLikePayload
  PostUnLike(input: PostUnLikeInput!): PostUnLikePayload
  CommentLike(input: CommentLikeInput!): CommentLikePayload
  CommentUnLike(input: CommentUnLikeInput!): CommentUnLikePayload
  PostCommentCreate(input: PostCommentCreateInput!): PostCommentCreatePayload
}

type UserLoginWithEmailPayload {
  token: String
  me: User
  error: String
  success: String
  clientMutationId: String
}

input UserLoginWithEmailInput {
  email: String!
  password: String!
  clientMutationId: String
}

type UserRegisterWithEmailPayload {
  token: String
  me: User
  error: String
  success: String
  clientMutationId: String
}

input UserRegisterWithEmailInput {
  name: String!
  email: String!
  password: String!
  clientMutationId: String
}

type PostCreatePayload {
  postEdge: PostEdge
  error: String
  success: String
  clientMutationId: String
}

input PostCreateInput {
  content: String!
  clientMutationId: String
}

type PostLikePayload {
  post: Post
  error: String
  success: String
  clientMutationId: String
}

input PostLikeInput {
  post: ID!
  clientMutationId: String
}

type PostUnLikePayload {
  post: Post
  error: String
  success: String
  clientMutationId: String
}

input PostUnLikeInput {
  post: ID!
  clientMutationId: String
}

type CommentLikePayload {
  comment: Comment
  error: String
  success: String
  clientMutationId: String
}

input CommentLikeInput {
  comment: ID!
  clientMutationId: String
}

type CommentUnLikePayload {
  comment: Comment
  error: String
  success: String
  clientMutationId: String
}

input CommentUnLikeInput {
  comment: ID!
  clientMutationId: String
}

type PostCommentCreatePayload {
  commentEdge: CommentEdge
  post: Post
  error: String
  success: String
  clientMutationId: String
}

input PostCommentCreateInput {
  post: ID!
  body: String!
  clientMutationId: String
}

type Subscription {
  PostNew(input: PostNewInput!): PostNewPayload
}

type PostNewPayload {
  post: Post
  clientSubscriptionId: String
}

input PostNewInput {
  clientSubscriptionId: String
}
