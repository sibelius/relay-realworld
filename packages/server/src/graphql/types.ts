import mongoose from 'mongoose';

import { DataLoaders } from '../modules/loader/loaderRegister';
import { IUser } from '../modules/user/UserModel';

declare type ObjectId = mongoose.Schema.Types.ObjectId;

export type GraphQLContext = {
  user?: IUser;
  dataloaders: DataLoaders;
};

export type LoaderFn = (ctx: GraphQLContext, id: string | ObjectId | object) => any;
