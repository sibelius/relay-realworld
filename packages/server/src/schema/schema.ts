import { GraphQLSchema } from 'graphql';

import QueryType from './QueryType';
import MutationType from './MutationType';
import SubscriptionType from './SubscriptionType';

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  subscription: SubscriptionType,
});
