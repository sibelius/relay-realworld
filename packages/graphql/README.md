# @world/graphql

GraphQL utilities to build graphql-js server faster and uniform
