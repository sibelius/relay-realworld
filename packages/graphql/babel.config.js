const config = require('@world/babel');

module.exports = config;
