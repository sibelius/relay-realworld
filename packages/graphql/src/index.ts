export { getObjectId } from './getObjectId';
export { createLoader, DataLoaderKey } from './createLoader';
export { connectionDefinitions, connectionArgs } from './connectionDefinitions';
export { mongooseIDResolver } from './mongooseIDResolver';
export { timestamps } from './timestampResolvers';
export { errorField } from './errorField';
export { successField } from './successField';
export { withFilter, ArgsWithFilter } from './withFilter';
