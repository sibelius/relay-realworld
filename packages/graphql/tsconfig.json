{
  "extends": "../../tsconfig.shared.json",
  "compilerOptions": {
    /* Redirect output structure to the directory. */
    "outDir": "./build/tsc",
    /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    "rootDir": "./src",
    /* Enable project compilation */
    "composite": true,
    /* Specify file to store incremental compilation information */
    "tsBuildInfoFile": "./build/tsconfig.tsbuildinfo"
  }
}
