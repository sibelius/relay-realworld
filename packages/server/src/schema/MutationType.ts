import { GraphQLObjectType } from 'graphql';

import UserMutations from '../modules/user/mutations';
import PostMutations from '../modules/post/mutations';
import LikeMutations from '../modules/like/mutations';
import CommentMutations from '../modules/comment/mutations';

const MutationType = new GraphQLObjectType({
  name: 'Mutation',
  fields: () => ({
    ...UserMutations,
    ...PostMutations,
    ...LikeMutations,
    ...CommentMutations,
  }),
});

export default MutationType;
