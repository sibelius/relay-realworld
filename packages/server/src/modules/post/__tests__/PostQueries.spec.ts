import { graphql } from 'graphql';

import { clearDbAndRestartCounters, connectMongoose, disconnectMongoose, sanitizeTestObject } from '../../../../test';
import { createUser } from '../../../modules/user/fixture/createUser';
import { schema } from '../../../schema/schema';
import { getContext } from '../../../getContext';
import { createPost } from '../fixture/createPost';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should all posts', async () => {
  const user = await createUser();
  const anotherUser = await createUser();

  await createPost({
    author: user,
  });

  await createPost({
    author: anotherUser,
  });

  // language=GraphQL
  const query = `
    query Q {
      posts(first: 10) {
        edges {
          node {
            content
            author {
              name
            }
          }
        }
      }
    }    
  `;

  const rootValue = {};
  const context = await getContext({ user });
  const variables = {};

  const result = await graphql(schema, query, rootValue, context, variables);
  expect(result.errors).toBeUndefined();

  console.log('result: ', result.data.posts);
  expect(result.data.posts.edges.length).toBe(2);
  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
