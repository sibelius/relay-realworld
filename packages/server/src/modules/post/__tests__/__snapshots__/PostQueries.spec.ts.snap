// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should all posts 1`] = `
Object {
  "posts": Object {
    "edges": Array [
      Object {
        "node": Object {
          "author": Object {
            "name": "user#2",
          },
          "content": "content#1",
        },
      },
      Object {
        "node": Object {
          "author": Object {
            "name": "user#1",
          },
          "content": "content#0",
        },
      },
    ],
  },
}
`;
